### PR TITLE
Measure components sharing a meter with siblings as a meter subtraction

### PR DESCRIPTION
**Pull request by shsms**
Components that share a parent meter with sibling meters or non-target
components had no meter fallback — an unreachable inverter degraded to
`COALESCE(#inv, 0)`. They are now measured as the parent meter(s) minus
the siblings, with the component readings as fallback.

## Changes

- New subtraction measurement: a target group covered by a parent meter
  next to sibling meters or measurable non-target components resolves to
  `COALESCE(..., #parent - #siblings, ...)`; generalized to partial
  groups (one unreachable inverter next to working siblings) and to
  diamonds (several parallel parent meters summed).
- The consumer formula measures its non-consumer components as one
  group, so siblings sharing a meter resolve to that meter once; fallback
  grid meters are excluded from subtraction (they carry the site's
  unmodeled consumer load).
- Stands-alone meter measurements stay total: the meter's reading is
  backed by its children's best-effort sum, so one offline meter no
  longer nulls a whole formula (grid meters stay bare).
- `prefer_meters_in_component_formulas` now defaults to `false`:
  component readings are the primary source and meters the fallback for
  the per-category formulas.

## Breaking

- `prefer_meters_in_component_formulas` defaults to `false`; set it to
  `true` to restore the meter-first source preference.

### CODE REVIEW

File: src/graph/formulas/fallback.rs [hunk 1]
```diff
@@ -154,9 +253,9 @@ fn meter_substitution<N: Node, E: Edge>(
     id: u64,
     targets: &BTreeSet<u64>,
 ) -> Result<Option<Substitution>, Error> {
-    if !is_measurable_component(graph.component(id)?, &graph.config) {
+    let Some(meters) = parent_meters(graph, id)? else {
```
Comment (llucax) [inline src/graph/formulas/fallback.rs:261]: 🔴 Asymmetric diamonds can drop a feed based on component ID order

`meter_substitution()` accepts the seed component’s parent-meter set as soon as every discovered sibling is targeted, but it does not verify that those siblings are fed only through the same set. For `M1 -> {A, B}` and `M2 -> B`, with targets `{A, B}`, processing lower-ID `A` first claims `M1` and removes `B` from `remaining`. The fallback then uses only `M1`; if `B` is processed first, the same topology correctly becomes a diamond over `M1 + M2`.

#### Impact
When component readings are unavailable, part of the group’s power is omitted. Formula correctness therefore changes solely when component IDs are reordered.

#### Related locations

- [`measurement_points()`](https://github.com/frequenz-floss/frequenz-microgrid-component-graph-rs/blob/a6d91f74c98f44d0b58466c4ab33868c257743a0/src/graph/formulas/fallback.rs#L167-L194): removes all siblings before a component with the complete parent set can be visited.
- [`fed_only_through()`](https://github.com/frequenz-floss/frequenz-microgrid-component-graph-rs/blob/a6d91f74c98f44d0b58466c4ab33868c257743a0/src/graph/formulas/fallback.rs#L305-L317): the required coverage check exists, but is only used by subtraction.

#### Suggested fix
Before accepting a substitution, verify that every covered component has compatible parent-meter coverage, or gather the union of parent meters for the complete sibling group before claiming it.

#### Suggested tests
Build the asymmetric topology twice with `A` and `B` IDs reversed. Both variants should resolve to the same diamond over `{M1, M2}` and generate equivalent formulas.

---

🔴 High Severity: very unlikely it is OK to merge without fixing first.

File: src/graph/formulas/fallback.rs [hunk 2]
```diff
@@ -173,9 +294,101 @@ fn meter_substitution<N: Node, E: Edge>(
     if meters.is_empty() {
         return Ok(None);
     }
-    Ok(Some(Substitution {
-        meters: meters.into_iter().collect(),
-        siblings,
+    for &meter in &meters {
+        if is_grid_meter(graph, graph.component(meter)?)? {
```
Comment (llucax) [inline src/graph/formulas/fallback.rs:318]: 🔴 Internal phantom loads are attributed to the target category

This guard rejects grid meters because their readings can contain unmodeled consumer load, but the configuration contract explicitly says that any meter with successors can carry such phantom loads. For an internal meter `M -> {PV, Battery}` that also carries unmodeled load `U`, a missing PV reading falls back to `M - Battery = PV + U`.

#### Impact
Consumer load can be reported as PV, battery, CHP, EV-charger, wind-turbine, or steam-boiler power. Depending on signs, this can overstate, understate, or reverse the category result exactly when the new fallback is selected.

#### Related locations

- [`ComponentGraphConfig`](https://github.com/frequenz-floss/frequenz-microgrid-component-graph-rs/blob/a6d91f74c98f44d0b58466c4ab33868c257743a0/src/config.rs#L27-L36): documents phantom loads on meters with successors generally, not only grid meters.
- [`subtraction_meters()`](https://github.com/frequenz-floss/frequenz-microgrid-component-graph-rs/blob/a6d91f74c98f44d0b58466c4ab33868c257743a0/src/graph/formulas/fallback.rs#L329-L392): treats a non-grid parent reading as exhaustive over modeled children.

#### Suggested fix
Require an explicit guarantee that a parent meter is exhaustive over the represented children before using subtraction. Without such metadata, conservatively skip subtraction for generic mixed meters or make this assumption an explicit opt-in.

#### Suggested tests
Evaluate a mixed internal meter with a known residual load and a missing target reading. The residual must not appear in the target category formula.

---

🔴 High Severity: very unlikely it is OK to merge without fixing first.

File: src/graph/formulas/fallback.rs [hunk 3]
```diff
@@ -186,15 +399,27 @@ fn measure<N: Node, E: Edge>(
...
-        return Ok(own);
+        // Standing alone means the meter's own reading is the only primary
+        // source, not that the term may go null with it: the children's
+        // best-effort sum still backs it, keeping the term total. A grid
+        // meter stays bare — it carries the site's unmodeled consumer load,
+        // which no sum of its children accounts for.
+        if is_grid_meter(graph, component)? {
+            return Ok(own);
+        }
+        return Ok(match sum(children.iter().map(|c| child_best_term(*c))) {
```
Comment (llucax) [inline src/graph/formulas/fallback.rs]: 🟡 An offline child meter still nulls the “total” fallback

The new stand-alone-meter fallback sums `child_best_term()`, but that helper emits child meters as bare `#id`. In the added mixed-meter topology the expression is `COALESCE(#2, COALESCE(#6, 0.0) + #3)`. If meters `#2` and `#3` are offline while their leaf components report, the sum is still null because `#3` has no fallback.

#### Impact
The consumer or grid formula can remain unavailable despite every leaf component reporting, contradicting this block’s stated guarantee that the term is total and the PR’s claim that one offline meter no longer nulls a whole formula.

#### Related locations

- [`child_best_term()`](https://github.com/frequenz-floss/frequenz-microgrid-component-graph-rs/blob/a6d91f74c98f44d0b58466c4ab33868c257743a0/src/graph/formulas/fallback.rs#L527-L534): returns a bare reference for child meters.
- [`test_consumer_formula_mixed_meter_with_component_submeter`](https://github.com/frequenz-floss/frequenz-microgrid-component-graph-rs/blob/a6d91f74c98f44d0b58466c4ab33868c257743a0/src/graph/formulas/generators/consumer.rs#L703-L742): its expected string exposes the null-propagating child term.

#### Suggested fix
Resolve child meters through `measure()` or an equivalent recursive helper so their own child fallback is preserved. Keep the existing special treatment for grid meters.

#### Suggested tests
Cover the state where the stand-alone parent meter and its child sub-meter are unavailable, but all leaf components report; the resulting formula must still evaluate to a value.

---

🟡 Medium Severity: probably best to fix before merging, OK to skip if there is a good reason.

File: RELEASE_NOTES.md [hunk 4]
```diff
@@ -8,12 +8,16 @@
...
 
 ## New Features
 
 - `ErrorKind` and `ValidationError` are now public. `Error::kind()` exposes the kind, and each `ValidationError` reports its `message()` and the `component_ids()` it involves, so individual validation failures (including detected cycles) can be inspected programmatically instead of parsed from a string.
 
+- Components that share a meter with sibling meters or components of another category (e.g. PV inverters next to a battery sub-meter under one "PV + battery" meter) now fall back to the parent meter minus those siblings (`COALESCE(..., #parent - #sub, ...)`) when their own readings are missing. This also applies to partial groups, so a single unreachable inverter can be measured as the meter minus its working siblings, and to diamonds, where a target behind several parallel meters is measured as their summed readings minus its non-target siblings.
```
Comment (llucax) [inline RELEASE_NOTES.md]: 🔵 Release notes omit the stand-alone-meter fallback change

The PR also changes a non-grid stand-alone meter from a bare `#meter` reference to `COALESCE(#meter, children...)`. That materially changes availability and null propagation in public formulas, but this entry only describes meter subtraction and diamonds.

#### Impact
Users upgrading cannot discover this additional formula behavior change from the release notes, even though generated formula strings and runtime fallback behavior change.

#### Suggested fix
Add a release-note item describing the child-reading fallback and the grid-meter exception.

---

🔵 Low Severity: addressing is optional (low impact / nice to have).

**Review by llucax — COMMENTED** [review]
🤖 **rAIview** (AI review here!)

The meter-subtraction design has two correctness issues that can undercount asymmetric diamonds or attribute unmodeled internal load to a component category. A nested offline sub-meter also defeats the new totality guarantee, and the release notes omit that stand-alone-meter behavior change.

File: src/graph/formulas/fallback.rs (hunk 1 above)
Comment (shsms) [reply src/graph/formulas/fallback.rs:261]: It is talking about a case that is not logical in terms of wiring.  We've never had support for these cases and we don't need to, but I'll see what I can do to keep it from flagging these again.

File: src/graph/formulas/fallback.rs (hunk 2 above)
Comment (shsms) [reply src/graph/formulas/fallback.rs:318]: > Require an explicit guarantee that a parent meter is exhaustive over the represented children before using subtraction.

we can't do that.  we're doing a best possible formula given the available information.  this is why we changed the default, making the meter subtraction the fallback from the sum of the inverters.  I think we can leave it at this.  but will add a comment, so it doesn't get flagged in the future.

File: src/graph/formulas/fallback.rs (hunk 1 above)
Comment (llucax) [reply src/graph/formulas/fallback.rs:261]: Yeah  i was about to ask if these reviews are being helpful or not  and I was thinking maybe we could add some extra context in a `AGENTS.md` file, but not sure how doable this could be. Let me know if you prefer if I skip the AI review of it's not being very effective.

**Comment by shsms** [comment]
Switching to draft, so that I can do another round of manual review and some testing tomorrow.

File: src/graph/formulas/fallback.rs (hunk 1 above)
Comment (shsms) [reply src/graph/formulas/fallback.rs:261]: Not necessary for us, but decided to implement anyway for completeness.  `meter_substitution()` now checks that the seed *and all siblings* are fed only through the claimed meter set. In the `M1 → {A, B}, M2 → B` case, the group now resolves through B with `{M1, M2}`.  Also add a test.

File: src/graph/formulas/fallback.rs (hunk 2 above)
Comment (shsms) [reply src/graph/formulas/fallback.rs:318]: added a comment.

File: src/graph/formulas/fallback.rs (hunk 3 above)
Comment (shsms) [reply src/graph/formulas/fallback.rs]: fixed, also renamed `child_best_term` to `child_best_effort_term`.

File: RELEASE_NOTES.md (hunk 4 above)
Comment (shsms) [reply RELEASE_NOTES.md]: done.

**Comment by shsms** [comment]
Wanna give it another look @llucax?

File: src/graph/formulas/fallback.rs (hunk 1 above)
Comment (llucax) [reply src/graph/formulas/fallback.rs:261]: > Yeah i was about to ask if these reviews are being helpful or not and I was thinking maybe we could add some extra context in a AGENTS.md file, but not sure how doable this could be. Let me know if you prefer if I skip the AI review of it's not being very effective.

What about this specifically?

File: src/graph/formulas/fallback.rs [hunk 5]
```diff
@@ -173,9 +311,196 @@ fn meter_substitution<N: Node, E: Edge>(
...
+/// overflowing the stack.
+fn reached_only_through_inner<N: Node, E: Edge>(
+    graph: &ComponentGraph<N, E>,
+    id: u64,
+    meters: &BTreeSet<u64>,
+    checked: &mut BTreeSet<u64>,
+) -> Result<bool, Error> {
+    for predecessor in graph.predecessors(id)? {
+        let predecessor_id = predecessor.component_id();
+        if meters.contains(&predecessor_id) || !checked.insert(predecessor_id) {
```
Comment (llucax) [inline src/graph/formulas/fallback.rs:353]: 🟡 Unconnected meter feeds pass the provenance guard

`reached_only_through_inner()` treats an already-visited predecessor as valid and returns `true` when a recursively visited meter has no predecessors. Consider `Grid -> GridMeter -> ParentMeter -> {A, B}` plus an unconnected `ExternalMeter -> B`, with `allow_unconnected_components(true)`. While building the formula for only `A`, the walk from `B` recurses into `ExternalMeter`; its empty predecessor list succeeds, so the code accepts `COALESCE(#A, #ParentMeter - #B, 0.0)`. But `#B` includes the external line while `#ParentMeter` does not, so this fallback undercounts `A`.

The same success-on-repeat rule accepts a closed off-root predecessor cycle as valid provenance. The added cycle test only checks termination, so it does not expose the resulting unsound formula.

#### Impact
Graphs explicitly permitted by `allow_unconnected_components` can generate materially incorrect category fallbacks. A long off-root predecessor chain also makes this recursive walk vulnerable to stack exhaustion during formula generation.

#### Related locations

- [`group_reached_only_through`](https://github.com/frequenz-floss/frequenz-microgrid-component-graph-rs/blob/4c8fbd33a0f156aaa30b3dc784e713c1f7f6a7eb/src/graph/formulas/fallback.rs#L368-L379): uses this result to qualify substitution and subtraction groups.
- [`test_off_root_cycle_terminates`](https://github.com/frequenz-floss/frequenz-microgrid-component-graph-rs/blob/4c8fbd33a0f156aaa30b3dc784e713c1f7f6a7eb/src/graph/formulas/fallback.rs#L2358-L2393): exercises termination but explicitly ignores formula soundness.

#### Suggested fix
Use an anchored, preferably iterative tri-state traversal. Every predecessor path should terminate at one of `meters`; a non-set meter with no predecessors and a back-edge should both reject the candidate rather than count as verified.

#### Suggested tests
Add the unconnected external-meter topology above and verify that a partial target does not use `#ParentMeter - #B`. Also assert a conservative direct measurement for an off-root cycle and cover a long predecessor chain without recursion.

---

🟡 Medium Severity: probably best to fix before merging, OK to skip if there is a good reason.

File: src/graph/formulas/generators/producer.rs [hunk 6]
```diff
@@ -170,8 +170,8 @@ mod tests {
                 "MIN(COALESCE(#6, #5, 0.0), 0.0) + ",
                 "MIN(COALESCE(#7, 0.0), 0.0) + ",
                 "MIN(COALESCE(#8, 0.0), 0.0) + ",
-                "MIN(COALESCE(#13, 0.0), 0.0) + ",
-                "MIN(COALESCE(#14, 0.0), 0.0)"
+                "MIN(COALESCE(#13, #12 - #14, 0.0), 0.0) + ",
```
Comment (llucax) [inline src/graph/formulas/generators/producer.rs:173]: 🟡 Group producer targets before resolving their shared meter

This expected formula resolves PV and CHP separately. For an internal meter `M -> {PV, CHP}`, that works while at least one device reports, but if both device readings are unavailable each difference (`#M - #CHP` and `#M - #PV`) is null and both terms fall through to `0.0`. The producer formula therefore reports zero even though `#M` measures the complete producer group.

The consumer path now avoids this exact limitation by passing all non-consumer targets to `aggregate_terms()` together. Grouping `{PV, CHP}` would likewise let the producer formula use `#M` when both child readings disappear, while retaining component-first behavior when they report.

#### Impact
A shared mixed producer meter can be available while `producer_formula()` reports no production. In the same telemetry state, `consumer_formula()` subtracts that meter successfully, so the generated formulas disagree about whether the production exists.

#### Related locations

- [`ProducerFormulaBuilder::build`](https://github.com/frequenz-floss/frequenz-microgrid-component-graph-rs/blob/4c8fbd33a0f156aaa30b3dc784e713c1f7f6a7eb/src/graph/formulas/generators/producer.rs#L37-L57): invokes `aggregate()` once per component, preventing the shared meter from covering the full producer target set.
- [`ConsumerFormulaBuilder::build_with_grid_meter`](https://github.com/frequenz-floss/frequenz-microgrid-component-graph-rs/blob/4c8fbd33a0f156aaa30b3dc784e713c1f7f6a7eb/src/graph/formulas/generators/consumer.rs#L165-L178): already resolves the complete non-consumer target set as groups.

#### Suggested fix
Collect the producer frontier into one target set, resolve it with `aggregate_terms()`, and apply the production clamp once per returned measurement group.

#### Suggested tests
For a mixed internal PV+CHP meter, evaluate the generated producer formula with both devices unavailable and the meter available. Also cover one missing device and both devices available to preserve the component-first preference.

---

🟡 Medium Severity: probably best to fix before merging, OK to skip if there is a good reason.

File: src/graph/formulas/fallback.rs [hunk 7]
```diff
@@ -173,9 +311,196 @@ fn meter_substitution<N: Node, E: Edge>(
...
+    // The subtracted siblings must measure only what flows through the parent
+    // meters, and must not lead to other targets (those are measured on their
+    // own, so subtracting them here would drop them from the total). A sibling
+    // that is itself one of the parent meters is rejected here too: the seed
+    // below it is always a nested target.
+    for &subtracted in &minus {
+        if !reached_only_through(graph, subtracted, &meters)? {
+            return Ok(None);
+        }
+        if reaches_any_below(graph, subtracted, targets)? {
```
Comment (llucax) [inline src/graph/formulas/fallback.rs:490]: 🟡 Cache descendant checks during subtraction discovery

Each subtraction candidate calls `reaches_any_below()` for every non-target sibling here and then again for overlap detection below. `measurement_points()` can retry the same candidate group from every target before finding a target whose wider parent set qualifies or rejects it.

A linear-sized valid topology can therefore trigger cubic work: put `n` targets and `n` sibling meters under one parent, let every sibling lead into the same `n`-node descendant chain, and give the highest-ID target an additional meter feed. For each of the first `n - 1` targets, subtraction walks the descendant chain twice for each sibling before `group_reached_only_through()` finally rejects the candidate because of the additional feed.

#### Impact
Formula generation can perform `O(n^3)` graph traversal on an `O(n)`-sized graph. Since formulas may be generated repeatedly and graph input can be large, this can cause severe CPU latency or an application-level denial of service.

#### Related locations

- [`measurement_points`](https://github.com/frequenz-floss/frequenz-microgrid-component-graph-rs/blob/4c8fbd33a0f156aaa30b3dc784e713c1f7f6a7eb/src/graph/formulas/fallback.rs#L166-L232): retries substitution/subtraction discovery for each remaining target.
- [`reaches_any_below`](https://github.com/frequenz-floss/frequenz-microgrid-component-graph-rs/blob/4c8fbd33a0f156aaa30b3dc784e713c1f7f6a7eb/src/graph/formulas/fallback.rs#L385-L398): starts a fresh full descendant traversal on every call.

#### Suggested fix
Precompute reachability to targets and subtraction candidates with reverse multi-source traversals, or cache these predicates and candidate results by node/parent-meter set for the duration of `measurement_points()`.

#### Suggested tests
Add an instrumented scaling test or benchmark for the topology above and assert that node/edge visits grow linearly or near-linearly rather than once per target-sibling pair.

---

🟡 Medium Severity: probably best to fix before merging, OK to skip if there is a good reason.

**Review by llucax — COMMENTED** [review]
🤖 **rAIview** (AI review here!)

The follow-up resolves the main issues from the previous review, but three medium-severity concerns remain: provenance checks can accept unanchored external feeds and generate unsound fallbacks, the producer formula misses a valid shared-meter fallback when all producer readings disappear, and subtraction discovery has a cubic worst case on a linear-sized graph.

File: src/graph/formulas/fallback.rs (hunk 1 above)
Comment (llucax) [reply src/graph/formulas/fallback.rs:261]: I just made a new AI round in the mean time, as you finally resolved most of the issues and didn't say explicitly to stop :)

File: src/graph/formulas/fallback.rs (hunk 5 above)
Comment (shsms) [reply src/graph/formulas/fallback.rs:353]: So this is the graph that triggers this case.  We need "allow unconnected components" to be true to trigger this.

In this case, we have an incomplete graph, and it looks like there's no power flowing through 1043, so from that perspective, the formula is still correct.

The fix is to disallow this case even when "allow unconnected components" is true.  Unconnected components should be completely disconnected from the grid-connected circuit.  Not just partially disconnected.  I'll make an issue.

Screenshot_2026-07-16_11-48-04

Comment (shsms) [reply src/graph/formulas/fallback.rs:353]: https://github.com/frequenz-floss/frequenz-microgrid-component-graph-rs/issues/45

File: src/graph/formulas/generators/producer.rs (hunk 6 above)
Comment (shsms) [reply src/graph/formulas/generators/producer.rs:173]: This is because targets are resolved one at a time so that we can wrap each of them in a MIN(), for example:  ` MIN(COALESCE(#13, #12 - #14, 0.0), 0.0) + MIN(COALESCE(#14, #12 - #13, 0.0), 0.0)`.

I already have a todo item to bring this up with Christoph, see if this would be ok, then come up with a plan to clean this up.

File: src/graph/formulas/fallback.rs (hunk 7 above)
Comment (shsms) [reply src/graph/formulas/fallback.rs:490]: The n^3 is possible when a meter has both inverters/chp/ev chargers and another meter with a subtree.  I can't imagine such subtrees having more than 10 components, so can't imagine this ever becoming a problem, but added a new commit with a fix anyway.

**Review by llucax — APPROVED** [review]

File: src/graph/formulas/fallback.rs (hunk 1 above)
Comment (shsms) [reply src/graph/formulas/fallback.rs:261]: I mean it is annoying, but they are not complete halucinations and it is nice to be made aware of the weird edge cases.  So I'd say it is fine.
